inc/fftw.hpp

```diff
@@ -0,0 +1,629 @@
+	: n_(0),
+	  n_padded_(0),
+	  extents_(),
+	  plan_(),
+	  data_(nullptr),
+	  data_transform_(nullptr),
+	  data_size_(0),
+	  data_transform_size_(0)
+      {
+	extents_ = interpret_as::column_major(cextents);
+	
+	n_ = std::accumulate(extents_.begin(),
+			     extents_.end(),
+			     1,
+			     std::multiplies<unsigned>());
+	if(Padding)
+	  n_padded_ = n_ / extents_[NDim-1] * (extents_[NDim-1]/2 + 1);
+
+	data_size_ = ( Padding ? 2*n_padded_ : n_ ) * sizeof(value_type);
+	data_transform_size_ = IsInplace ? 0 : n_ * sizeof(ComplexType);
+
+	traits::thread_api<TPrecision>::init_threads();
+	traits::thread_api<TPrecision>::plan_with_threads(-1);
+      }
+
+      ~FftwImpl(){
+	traits::thread_api<TPrecision>::cleanup_threads();
+      }
+      
+      /**
+       * Returns allocated memory for FFT
+       */
+      size_t getAllocSize() {
+	return data_size_ + data_transform_size_;
+      }
+
+      // create FFT plan handle
+      void init_forward() {
+
+	//NOTE: contents of data_ et al are overwritten by this call
+	plan_ = traits::plan<TPrecision>::create(extents_, data_, data_transform_, traits::forward);
+      }
+
+      // recreates plan if needed
+      void init_backward() {
+
+	traits::plan<TPrecision>::destroy(plan_);
+
+	//NOTE: contents of data_ et al are overwritten by this call
+	plan_ = traits::plan<TPrecision>::create(extents_, data_, data_transform_, traits::backward);
+      
+      }
+
+      /**
+       * Returns estimated allocated memory on device for FFT plan, accessing the plan size in fftw is impossible
+       
+       */
+      size_t getPlanSize() {
+      
+	if(IsInplace)
+	  return data_size_;
+	else
+	  return data_size_ + data_transform_size_;
+	
+      }
+
+      //////////////////////////////////////////////////////////////////////////////////////
+      // --- next methods are benchmarked ---
+
+      void malloc() {
+
+	data_ = (value_type*)traits::memory_api<TPrecision>::malloc(data_size_);
+      
+	if(IsInplace){
+	  data_transform_ = reinterpret_cast<ComplexType*>(data_);
+	}
+	else{
+	  data_transform_ = (ComplexType*)traits::memory_api<TPrecision>::malloc(data_transform_size_);
+	}
+		   
+      }
+
+
+      void execute_forward() {
+	traits::plan<TPrecision>::execute(plan_);
+      }
+
+      void execute_backward() {
+	traits::plan<TPrecision>::execute(plan_);
+      }
+
+      template<typename THostData>
+      void upload(THostData* input) {
+
+	//TODO: unnecessary?
+      
+	// if(Padding) // real + inplace + ndim>1
+	// {
+	//   size_t w      = extents_[NDim-1] * sizeof(THostData);
+	//   size_t h      = n_ * sizeof(THostData) / w;
+	//   size_t pitch  = (extents_[NDim-1]/2+1) * sizeof(ComplexType);
+	//   CHECK_CUDA(cudaMemcpy2D(data_, pitch, input, w, w, h, cudaMemcpyHostToDevice);
+	// }else{
+	//   CHECK_CUDA(cudaMemcpy(data_, input, data_size_, cudaMemcpyHostToDevice);
+	// }
+
+      
+      }
+
+      template<typename THostData>
+      void download(THostData* output) {
+
```

Comment:
copy computed data to output.
(If nothing happens to output and input pointer (which are actually the same), then there are no errors, thus Boost test is fine.)